chat-android/src/main/java/com/ably/chat/AtomicCoroutineScope.kt

```diff
@@ -0,0 +1,94 @@
+ * See [Kotlin Dispatchers](https://kt.academy/article/cc-dispatchers) for more information.
+ */
+class AtomicCoroutineScope(private val scope: CoroutineScope = CoroutineScope(Dispatchers.Default)) {
+
+    private val sequentialScope: CoroutineScope = CoroutineScope(Dispatchers.Default.limitedParallelism(1))
+
+    private class Job(
+        private val priority: Int,
+        val coroutineBlock: suspend CoroutineScope.() -> Any,
+        val deferredResult: CompletableDeferred<Any>,
+        val queuedPriority: Int,
+    ) :
+        Comparable<Job> {
+        override fun compareTo(other: Job): Int {
+            if (this.priority == other.priority) {
+                return this.queuedPriority.compareTo(other.queuedPriority)
+            }
+            return this.priority.compareTo(other.priority)
+        }
+    }
+
+    private val jobs: PriorityBlockingQueue<Job> = PriorityBlockingQueue() // Accessed from both sequentialScope and async method
+    private var isRunning = false // Only accessed from sequentialScope
+    private var queueCounter = 0 // Only accessed from synchronized method
+
+    val finishedProcessing: Boolean
+        get() = jobs.isEmpty() && !isRunning
+
+    val pendingJobCount: Int
+        get() = jobs.count()
+
+    /**
+     * Defines priority for the operation execution and
+     * executes given coroutineBlock mutually exclusive under given scope.
+     */
+    @Synchronized
+    fun <T : Any>async(priority: Int = 0, coroutineBlock: suspend CoroutineScope.() -> T): CompletableDeferred<T> {
+        val deferredResult = CompletableDeferred<Any>()
+        jobs.add(Job(priority, coroutineBlock, deferredResult, queueCounter++))
+        sequentialScope.launch {
+            if (!isRunning) {
+                isRunning = true
+                while (jobs.isNotEmpty()) {
+                    val job = jobs.poll()
+                    job?.let {
+                        safeExecute(it)
+                    }
+                }
+                isRunning = false
+            }
+        }
+
+        @Suppress("UNCHECKED_CAST")
+        return deferredResult as CompletableDeferred<T>
+    }
+
+    private suspend fun safeExecute(job: Job) {
+        try {
+            scope.launch {
+                try {
+                    val result = job.coroutineBlock(this)
+                    job.deferredResult.complete(result)
+                } catch (t: Throwable) {
+                    job.deferredResult.completeExceptionally(t)
+                }
+            }.join()
+        } catch (t: Throwable) {
+            job.deferredResult.completeExceptionally(t)
+        }
+    }
+
+    /**
+     * Cancels ongoing and pending operations with given error.
+     */
+    @Synchronized
+    fun cancel(message: String?, cause: Throwable? = null) {
+        queueCounter = 0
+        sequentialScope.coroutineContext.cancelChildren(CancellationException(message, cause))
+    }
```

Comment:
Actually, `sequentialScope.coroutineContext.cancelChildren` cause all children job to cancel. That means `CancellationException` will be thrown at suspension point. We already have a mechanism to catch this exception inside `safeExecute` method. This will make sure all queued jobs are processed by while loop, and while processing each job, exception will be thrown and will be caught, setting it as a result for the job. At the end of while loop `isRunning` will be set to false. You might like to reach about children job cancellation here -> https://kt.academy/article/cc-cancellation#cancellation-in-a-coroutine-scope. 
Also, we have a working test to have this expected behavior -> https://github.com/ably-labs/ably-chat-kotlin/blob/290fb84dc0fbf0c70236f75c722bf057c9d27510/chat-android/src/test/java/com/ably/chat/AtomicCoroutineScopeTest.kt#L252